### ISSUE TITLE
Openshift: Using 'enableIpAddressResolution' in parameters.config not working

### ISSUE THREAD

**Issue by atoy3731** [DecipherNow/gm-fabric-jvm#112]
**Service Info:**
Name: apps
Version:  1.1.2-1.201802051842
GM Fabric Version:  0.2.1

**Issue:**
I have this set in `parameters.config`:

    -com.deciphernow.server.config.ipAddress.enableIpAddressResolution=true
    -com.deciphernow.server.config.ipAddress.useNetworkInterfaceName=eth0 

This should take the eth0 IP address and announce it to ZK but it is still announcing the container hostname:

    [zk: localhost:2181(CONNECTED) 6] get /services/apps/1.1/https/member_0000000010
    {"serviceEndpoint":{"host":"apps-1-1-1-88rfw","port":9779},"additionalEndpoints":{},"status":"ALIVE","shard":0}

EDIT: This also affects using `-com.deciphernow.server.config.os.env.hostname` and setting the environment variable equal to an IP address. The framework will do a DNS lookup on the IP address and still announce the hostname to ZK.

**Comment by joshua-rutherford** [comment]
I can confirm this by building the tutorial project and running it with:
```
$ export OPENSHIFT_IP="172.217.7.23"
$ /usr/bin/java \
  -classpath "./server/target/exemplar-server-0.1.0-SNAPSHOT-app/exemplar-server-0.1.0-SNAPSHOT/lib/*" \
  -Xms48m \
  -Xmx96m \
  -Dcom.deciphernow.server.config.rest.httpPort=:9679 \
  -Dcom.deciphernow.server.config.rest.httpsPort=:9779 \
  -Dcom.deciphernow.server.config.thrift.port=:9879 \
  -Dcom.deciphernow.server.config.admin.port=:9979 \
  -Dcom.deciphernow.server.config.tls.keyStore=./localhost.jks \
  -Dcom.deciphernow.server.config.tls.keyStorePass=password \
  -Dcom.deciphernow.server.config.tls.trustStore=./truststore.jks \
  -Dcom.deciphernow.server.config.tls.trustStorePass=password \
  -Dcom.deciphernow.server.config.filter.blockHttp=True \
  -Dcom.example.exemplar.acl.whitelist=./accessControlList.txt \
  -Dcom.deciphernow.server.config.zk.announcementPoint=/services/exemplar-service/0.1.0 \
  -Dcom.deciphernow.server.config.zk.zookeeperConnection=127.0.0.1:2181 \
  -Dcom.deciphernow.server.config.os.env.hostname=OPENSHIFT_IP \
  com.example.exemplar.exemplar
```
Regardless of the settings for `-com.deciphernow.server.config.ipAddress.enableIpAddressResolution=true` I end announcing as one of Googles servers (`lga25s56-in-f23.1e100.net`).

**Comment by joshua-rutherford** [comment]
Note that this does not occur if I use a) an IP address without a DNS entry or b) and hostname.

**Comment by ghershfield** [comment]
Exactly what I was saying about the DNS issue

**Comment by atoy3731** [comment]
@ghershfield Is there a COA for a potential fix?

**Comment by joshua-rutherford** [comment]
Also, note that the above system properties reflect a change I made from the tutorial to accept the acl list from system properties (there is no `com.example.exemplar.acl.whitelist` flag in the normal service).

**Comment by atoy3731** [comment]
Is this issue being worked? It's a blocker for getting these services working in ADX.

**Comment by mikejlong60** [comment]
I am working on this now.  I hope to finish it today.

**Comment by mikejlong60** [comment]
Update on status:
The Framework is never loading these properties from the parameters.config file or from environment variables.  
-com.deciphernow.server.config.ipAddress.enableIpAddressResolution=true
-com.deciphernow.server.config.ipAddress.useNetworkInterfaceName=eth0 

For example if com.deciphernow.server.config.ipAddress.enableIpAddressResolution is hardwired to be true,  you get proper assignment of an IP address to the hostname field for ZK registration.  I am currently working on understanding what part of the mechanism is missing to load these properties.  It used to work.

**Comment by joshua-rutherford** [comment]
All, I've been looking at this most of the day and I'm afraid that our options here are fairly limited.  The location where all this happens is here: https://github.com/twitter/commons/blob/master/src/java/com/twitter/common/zookeeper/ServerSets.java#L133.  On that line the underlying implementation invokes `InetSocketAddress.getHostName` which according to the documentation:

> Note: This method may trigger a name service reverse lookup if the address was created with a literal IP address.

As a result, I've looked into the easiest options to override this:

1. Override this by wrapping an `InetSocketAddress` with our own implementation that overrides the `getHostName` method to conditionally return the address or hostname depending upon how it is constructed.  Unfortunately, `InetSocketAddress.getHostName` method as final so I cannot override it.
2. Looking at the OpenJDK implementation, the `InetSocketAddress.getHostName` method is just passed off to `InetAddress.getHostName` so I looked at doing a similar thing there.  Unfortunately, `InetAddress` is not an interface, but a class and it's constructor is package private (as are it's descendants, both of which are final).  We cannot sneak this past the package private constructor by putting our implementation in `java.net` since that package is protected for security reasons.
3. Java has an extensible name service provider capability which we have exploited in tests before (i.e., https://github.com/DecipherNow/moby/tree/master/moby-dns).  I've never used this approach in production but here is one that is getting the test to register one of Google's servers by address not hostname.
4. We can do the announcement ourselves by implementing a custom `Announcer`.  This would likely boil down to us largely copying the `ZkAnnouncer`, `ServerSet`, `ServerSetImpl` and `ServerSets` and a changing only what need to be changed to make our implementation of `ServerSets.toEndpoint` not do the blind call to `InetSocketAddress.getHostName` in the cases we want. 

Personally, the easiest by far would be 1 or 2, but those are really not possible without really mucking around with things via reflection and altering security.  The next easiest is 3 and I've provided an implementation that works here: https://github.com/DecipherNow/gm-fabric-jvm/pull/125.  4 is doable with some effort but will require a good bit more code (mostly copy an paste though).  Still, I like how targeted my solution in 3 is (i.e., I only muck with reverse lookups for a single IP address).
